####################
## Hybrid Assembly Pipeline 3: 
## Unicycler with existing long assembly
####################

rule all_draft3:
 input:
  expand(config['project_dir'] + '/reports/{step}/{barcode}/asm.aln.tsv', 
    step=['09_unicycler_long'], barcode=BARCODES)

rule run_unicycler_existing_long:
 input:
  reads = config['project_dir'] + '/03_subsampled_reads/{barcode}/reads.fastq.gz',
  R1 = config['project_dir'] + '/' + config['short_reads_fp'] + '/{barcode}/R1.fastq',
  R2 = config['project_dir'] + '/' + config['short_reads_fp'] + '/{barcode}/R2.fastq',
  longasm = config['project_dir'] + '/05_nanopolish/{barcode}/nanopolish.contigs.fasta'
 output:
  contigs = config['project_dir'] + '/09_unicycler_long/{barcode}/assembly.fasta'
 threads: 16
 params:
  outdir = config['project_dir'] + '/09_unicycler_long/{barcode}',
  merge_mode = config['merge_mode']
 shell:
  """
  unicycler -1 {input.R1} -2 {input.R2} -l {input.reads} -o {params.outdir} -t {threads} --existing_long_read_assembly {input.longasm} --mode {params.merge_mode}
  """

rule assess_unicycler_existing_long:
 input:
  genome = config['project_dir'] + '/' + config['genome_fp'],
  draft_v3 = config['project_dir'] + '/09_unicycler_long/{barcode}/assembly.fasta',
  python_chop_fp = config['basecalling_cmp_fp'] + '/chop_up_assembly.py',
  python_ident_fp = config['basecalling_cmp_fp'] + '/read_length_identity.py'
 output:
  asm_pieces = config['project_dir'] + '/09_unicycler_long/{barcode}/asm_pieces.fasta',
  asm_aln = config['project_dir'] + '/09_unicycler_long/{barcode}/asm.aln.paf',
  asm_table = config['project_dir'] + '/reports/09_unicycler_long/{barcode}/asm.aln.tsv'
 threads: 8
 shell:
  """
  python {input.python_chop_fp} {input.draft_v3} 10000 > {output.asm_pieces}
  minimap2 -x map10k -t {threads} -c {input.genome} {output.asm_pieces} > {output.asm_aln}
  python {input.python_ident_fp} {output.asm_pieces} {output.asm_aln} > {output.asm_table}
  """
